internal/provider/kubernetes/tlsroute.go

```diff
@@ -0,0 +1,323 @@
+				if err := r.client.Get(ctx, svcKey, svc); err != nil {
+					if errors.IsNotFound(err) {
+						// The ref's service doesn't exist in the cache, so remove it from
+						// the resource map if it exists.
+						if _, ok := r.resources.Services.Load(svcKey); ok {
+							r.resources.Services.Delete(svcKey)
+							log.Info("deleted service from resource map")
+						}
+					}
+					return reconcile.Result{}, fmt.Errorf("failed to get service %s/%s",
+						svcKey.Namespace, svcKey.Name)
+				}
+
+				// The backendRef Service exists, so add it to the resource map.
+				r.resources.Services.Store(svcKey, svc)
+				log.Info("added service to resource map")
+			}
+		}
+	}
+
+	if !found {
+		// Delete the tlsroute from the resource map.
+		r.resources.TLSRoutes.Delete(request.NamespacedName)
+		log.Info("deleted tlsroute from resource map")
+
+		// Delete the Namespace and Service from the resource maps if no other
+		// routes exist in the namespace.
+		routeList = &gwapiv1a2.TLSRouteList{}
+		if err := r.client.List(ctx, routeList, &client.ListOptions{Namespace: request.Namespace}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("error listing tlsroutes")
+		}
+		if len(routeList.Items) == 0 {
+			r.resources.Namespaces.Delete(request.Namespace)
+			log.Info("deleted namespace from resource map")
+			r.resources.Services.Delete(request.NamespacedName)
+			log.Info("deleted service from resource map")
+		}
+	}
+
+	log.Info("reconciled tlsroute")
+
+	return reconcile.Result{}, nil
+}
+
+// validateTLSRouteBackendRef validates that ref is a reference to a local Service.
+func validateTLSRouteBackendRef(ref *gwapiv1a2.BackendRef) error {
+	switch {
+	case ref == nil:
+		return nil
+	case ref.Group != nil && *ref.Group != corev1.GroupName:
+		return fmt.Errorf("invalid group; must be nil or empty string")
+	case ref.Kind != nil && *ref.Kind != gatewayapi.KindService:
+		return fmt.Errorf("invalid kind %q; must be %q",
+			*ref.BackendObjectReference.Kind, gatewayapi.KindService)
+	case ref.Namespace != nil:
+		return fmt.Errorf("invalid namespace; must be nil")
+	}
+
+	return nil
+}
+
+// subscribeAndUpdateStatus subscribes to tlsroute status updates and writes it into the
+// Kubernetes API Server
+func (r *tlsRouteReconciler) subscribeAndUpdateStatus(ctx context.Context) {
+	// Subscribe to resources
+	for snapshot := range r.resources.TLSRouteStatuses.Subscribe(ctx) {
+		r.log.Info("received a status notification")
+		updates := snapshot.Updates
+		for _, update := range updates {
+			// skip delete updates.
+			if update.Delete {
+				continue
+			}
+			key := update.Key
+			val := update.Value
+			r.statusUpdater.Send(status.Update{
+				NamespacedName: key,
+				Resource:       new(gwapiv1a2.TLSRoute),
+				Mutator: status.MutatorFunc(func(obj client.Object) client.Object {
+					if _, ok := obj.(*gwapiv1a2.TLSRoute); !ok {
+						panic(fmt.Sprintf("unsupported object type %T", obj))
+					}
+					return val
```

Comment:
recommend only updating the required field similar to https://github.com/envoyproxy/gateway/blob/34afca0f0ad17e4caed3cda7a1ac2a1a1658ab11/internal/provider/kubernetes/httproute.go#L338